src/actor/Verifier.ts

```diff
@@ -0,0 +1,148 @@
+   * @returns A session and a message object.
+   * The **session** object will be used in [[verifyPresentation]] and should be kept private by the verifier.
+   * The **message** object should be sent to the Claimer and used in [[createPresentation]].
+   */
+  public async finalize(
+    allowPE: boolean
+  ): Promise<[gabi.CombinedVerificationSession, IRequestClaimsForCTypes]> {
+    const { session, message } = await this.builder.finalise()
+    return [
+      session,
+      {
+        type: MessageBodyType.REQUEST_CLAIMS_FOR_CTYPES,
+        content: {
+          ctypes: this.ctypes,
+          peRequest: message,
+          allowPE,
+        },
+      },
+    ]
+  }
+}
+
+/**
+ * Initiates a verification by creating a request on the Verifier's side.
+ *
+ * @returns A [[PresentationRequestBuilder]] based on a [[CType]] and a list of required disclosed attributes of the [[Credential]]s.
+ */
+export function newRequest(): PresentationRequestBuilder {
+  return new PresentationRequestBuilder()
+}
+
+/**
+ * [ASYNC] Verifies the Claimer's presentation of [[Credential]]s.
+ *
+ * @param presentation The Claimer's presentation of the [[Credential]]s that should be verified, the result of [[createPresentation]].
+ * @param session The Verifier's private verification session created in [[finalize]].
+ * @param latestAccumulators The list of the latest accumulators for each [[Attester]] which signed a [[Credential]] of this presentation.
+ * @param attesterPubKeys The privacy enhanced public keys of all [[AttesterIdentity]]s which signed the [[Credential]]s.
+ * @returns An array representing
+ * **at index 0** whether the [[Credential]]s could be verified
+ * and **at index 1** an array of [[Claim]]s restricted on the disclosed attributes selected in [[requestPresentationForCtype]].
+ */
+export async function verifyPresentation(
+  presentation: ISubmitClaimsForCTypes,
+  session?: gabi.CombinedVerificationSession,
+  latestAccumulators?: gabi.Accumulator[],
+  attesterPubKeys?: PublicAttesterIdentity[]
+): Promise<[boolean, any[]]> {
+  if (presentation.type === MessageBodyType.SUBMIT_CLAIMS_FOR_CTYPES_PUBLIC) {
+    const attestedClaims = presentation.content.map(
+      AttestedClaim.fromAttestedClaim
+    )
+    const allVerified = await Promise.all(attestedClaims.map(ac => ac.verify()))
+    return [allVerified.every(b => b), attestedClaims]
+  }
+  if (presentation.type === MessageBodyType.SUBMIT_CLAIMS_FOR_CTYPES_PE) {
+    if (
+      typeof session === 'undefined' ||
+      typeof latestAccumulators === 'undefined' ||
+      typeof attesterPubKeys === 'undefined'
+    ) {
+      throw new Error(`Received privacy enhanced presentation. Require:
+      Session: ${session}
+      accumulators: ${latestAccumulators}
+      public keys: ${attesterPubKeys}`)
+    }
+    const { verified, claims } = await gabi.Verifier.verifyCombinedPresentation(
```

Comment:
Did we get all properties, we requested?

Reply:
fixed in 35168366c4c4b12a47b582045c7103cfa64fcf70

Reply:
We would have to go over this together again. Not sure, what I meant here anymore.

Reply:
The verifier would request specific properties. We check if all requested properties are provided by the claimer, but only  for the PE version. We also need to check that for the public presentation.